File: src/main/java/com/hatzy/refacto/hatzyRefacto/Yatzy.java
package com.hatzy.refacto.hatzyRefacto;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static java.util.Arrays.asList;

public class Yatzy {

    public static int chance(int... dice) {
        return sum(dice);
    }

    public static int yatzy(int... dice) {
        if (diceFrequencies(dice).containsValue(5)) {
            return 50;
        }
        return 0;
    }

    public static int ones(int... dice) {
        return numberOfDiceFrequency(1, dice);
    }

    public static int twos(int... dice) {
        return numberOfDiceFrequency(2, dice);
    }

    public static int threes(int... dice) {
        return numberOfDiceFrequency(3, dice);
    }

    public static int fours(int... dice) {
        return numberOfDiceFrequency(4, dice);
    }

    public static int fives(int... dice) {
        return numberOfDiceFrequency(5, dice);
    }

    public static int sixes(int... dice) {
        return numberOfDiceFrequency(6, dice);
    }

    public static int two_pair(int... dice) {
        Map<Integer, Integer> frequencies = diceFrequencies(dice);
        int score = 0;
        boolean containsTwoPair = frequencies
                .values()
                .stream()
                .filter(value -> value >= 2)
                .count() == 2;

        if (containsTwoPair) {
            for (int maxFace : asList(6, 5, 4, 3, 2, 1)) {
                if (frequencies.get(maxFace) >= 2) {
                    score += maxFace * 2;
                }
            }
        }
        return score;
    }

    public static int score_pair(int... dice) {
        return numberOfAKind(2, dice);
    }

    public static int four_of_a_kind(int... dice) {
        return numberOfAKind(4, dice);
    }

    public static int three_of_a_kind(int... dice) {
        return numberOfAKind(3, dice);
    }

    public static int smallStraight(int... dice) {
        if (isStraight(dice) && diceFrequencies(dice).get(6) == 0) {
            return sum(dice);
        }
        return 0;
    }

    public static int largeStraight(int... dice) {
        if (isStraight(dice) && diceFrequencies(dice).get(1) == 0) {
            return sum(dice);
        }
        return 0;
    }

    public static int fullHouse(int... dice) {
        Map<Integer, Integer> frequencies = diceFrequencies(dice);
        if (frequencies.containsValue(2) && frequencies.containsValue(3)) {
            return sum(dice);
        }
        return 0;
    }

    private static Map<Integer, Integer> diceFrequencies(int... dice) {
        HashMap<Integer, Integer> frequencies = new HashMap<>();
        for (int face : asList(1, 2, 3, 4, 5, 6)) {
            frequencies.put(face, 0);
        }
        for (int die : dice) {
            frequencies.put(die, frequencies.get(die) + 1);
        }

        return frequencies;
    }

    private static int numberOfDiceFrequency(int number, int... dice) {
        return diceFrequencies(dice).get(number) * number;
    }

    private static int numberOfAKind(int number, int... dice) {
        for (int maxFace : asList(6, 5, 4, 3, 2, 1)) {
            if (diceFrequencies(dice).get(maxFace) >= number) {
                return maxFace * number;
            }
        }
        return 0;
    }

    private static boolean isStraight(int... dice) {
        return diceFrequencies(dice)
                .values()
                .stream()
                .filter(value -> value == 1)
                .count() == 5;
    }

    private static int sum(int[] dice) {
        return Arrays
                .stream(dice)
                .sum();
    }
}

File: javascript/lib/yatzy.js
var Yatzy = function() {}

Yatzy.chance = function(dice) {
    return Yatzy.sumDice(dice);
}

Yatzy.yatzy = function(dice) {
    var frequencies = Yatzy.diceFrequencies(dice);
    var values = Object.keys(frequencies).map(function(key) {
                        return frequencies[key];
                    });

    return values.find((n) => n === 5) ? 50 : 0;
}

Yatzy.ones = function(dice) {
    return Yatzy.numberOfDiceFrequency(1, dice);
}

Yatzy.twos = function(dice) {
    return Yatzy.numberOfDiceFrequency(2, dice);
}

Yatzy.threes = function(dice) {
    return Yatzy.numberOfDiceFrequency(3, dice);
}

Yatzy.fours = function(dice) {
    return Yatzy.numberOfDiceFrequency(4, dice);
}

Yatzy.fives = function(dice) {
    return Yatzy.numberOfDiceFrequency(5, dice);
}

Yatzy.sixes = function(dice) {
    return Yatzy.numberOfDiceFrequency(6, dice);
}

Yatzy.score_pair = function(dice)
{
    return Yatzy.numberOfAKind(2, dice);
}

Yatzy.two_pair = function(dice)
{
    let score = 0;
    var frequencies = Yatzy.diceFrequencies(dice);
    var values = Object.keys(frequencies).map(function(key) {
                        return frequencies[key];
                    });
        if (values.filter( (n) => n >= 2).length === 2) {
            [6, 5, 4, 3, 2, 1].map( (dice_value) => {
                if (frequencies[dice_value] >= 2)
                    score += 2*dice_value;
            } );
        }
        return score;
}

Yatzy.four_of_a_kind = function(dice)
{
    return Yatzy.numberOfAKind(4, dice);
}

Yatzy.three_of_a_kind = function(dice)
{
    return Yatzy.numberOfAKind(3, dice);
}

Yatzy.smallStraight = function(dice)
{
    return (Yatzy.isStraight(dice) && Yatzy.diceFrequencies(dice)[6] === 0) ? 15 : 0;
}

Yatzy.largeStraight = function(dice)
{
    return (Yatzy.isStraight(dice) && Yatzy.diceFrequencies(dice)[6] === 1) ? 20 : 0;
}

Yatzy.fullHouse = function(dice)
{
    const frequencies = Yatzy.diceFrequencies(dice);
    var values = Object.keys(frequencies).map(function(key) {
                        return frequencies[key];
                    });
    if (values.find( (n) => n === 3) 
            && values.find( (n) => n === 2))
       return Yatzy.sumDice(dice);
    else
       return 0;
}

Yatzy.diceFrequencies = function(dice){
    let facesCount = {1:0, 2:0, 3:0, 4:0, 5:0, 6:0};
    dice.forEach(number => facesCount[number]++);
    return facesCount;
}

Yatzy.numberOfDiceFrequency = function(number, dice){
    return Yatzy.diceFrequencies(dice)[number]*number;
}

Yatzy.numberOfAKind = function(number, dice){
    const frequencies = Yatzy.diceFrequencies(dice);

    let scores = [6, 5, 4, 3, 2, 1].map( (maxFace) => {
        if (frequencies[maxFace] >= number)
            return number*maxFace;
        else
            return 0;
    } );

    scores = scores.filter( (score) => score > 0);

    return scores.length > 0 ? scores[0] : 0;
}

Yatzy.isStraight = function(dice){
    var frequencies = Yatzy.diceFrequencies(dice);
    var values = Object.keys(frequencies).map(function(key) {
                        return frequencies[key];
                    });

    return values.filter( (value) => value === 1).length === 5;
}

Yatzy.sumDice = function(dice){
    let sumValues = 0;
    dice.forEach(value => sumValues += value);
    return sumValues;
}

module.exports = Yatzy;